ArduCopter/mode_zigzag.cpp

```diff
@@ -0,0 +1,373 @@
+        // initialize's loiter position and velocity on xy-axes from current pos and velocity
+        wp_nav->init_loiter_target();
+
+        // initialize vertical speed and acceleration's range
+        pos_control->set_speed_z(-g.k_param_pilot_speed_up, g.k_param_pilot_speed_up);
+        // the parameters are maximum climb and descent rates
+        pos_control->set_accel_z(g.pilot_accel_z);
+        // Global parameters are all contained within the 'g' class.
+        // Parameters g;
+
+        // initialise position_z and desired velocity_z
+        if (!pos_control->is_active_z()) {
+            // is_active_z - returns true if the z-axis position controller has been run very recently
+            pos_control->set_alt_target_to_current_alt();
+            pos_control->set_desired_velocity_z(inertial_nav.get_velocity_z());
+        }
+
+        // initialise waypoint state
+        zigzag_waypoint_state.A_hasbeen_defined = false;
+        zigzag_waypoint_state.B_hasbeen_defined = false;
+        in_zigzag_manual_control = true;
+        zigzag_is_between_A_and_B = false;
+        zigzag_judge_moving.is_keeping_time = false;
+
+        return true;
+    }
+    else {
+        return false;
+    }
+}
+
+// zigzag_run - runs the zigzag controller
+// should be called at 100hz or more
+void Copter::ModeZigzag::run()
+{
+    // if not auto armed or motors not enabled set throttle to zero and exit immediately
+    if (!motors->armed() || !ap.auto_armed || !motors->get_interlock() || ap.land_complete) {
+#if FRAME_CONFIG == HELI_FRAME  // Helicopters always stabilize roll/pitch/yaw
+        // call attitude controller
+        attitude_control->input_euler_angle_roll_pitch_euler_rate_yaw(0, 0, 0, get_smoothing_gain());
+        attitude_control->set_throttle_out(0, false, g.throttle_filt);
+#else
+        motors->set_desired_spool_state(AP_Motors::DESIRED_SPIN_WHEN_ARMED);
+        // multicopters do not stabilize roll/pitch/yaw when disarmed
+        attitude_control->set_throttle_out_unstabilized(0, true, g.throttle_filt);
+#endif
+        return;
+    }
+
+    // if B has not been defined
+    if (!zigzag_waypoint_state.B_hasbeen_defined) {
+        // receive pilot's inputs, do position and attitude control
+        zigzag_manual_control();
+    }
+
+    //else, if B has already been defined
+    //if it's in manual control part
+    //receive pilot's inputs, do position and attitude control
+    //else
+    //judge if the plane has arrived at the current destination
+    //if yes, go to the manual control part by modifying parameter
+    //else, fly to current destination
+    else {
+        if (in_zigzag_manual_control) {
+
+            zigzag_manual_control();
+        }
+        else { //auto flight
+
+            if (zigzag_has_arr_at_dest()) {  //if the plane has arrived at the current destination
+                in_zigzag_manual_control = true;
+                wp_nav->init_loiter_target();
+            }
+            else {
```

Comment:
I think that the way of writing the else sentence is different from the coding style.